syntax = "proto2";
import 'ostorlab/agent/message/proto/v3/asset/domain_name/domain_name.proto';
import 'ostorlab/agent/message/proto/v3/asset/ip/v4/v4.proto';
import 'ostorlab/agent/message/proto/v3/asset/ip/v6/v6.proto';
import 'ostorlab/agent/message/proto/v3/asset/link/link.proto';
import 'ostorlab/agent/message/proto/v3/asset/store/android_store/android_store.proto';
import 'ostorlab/agent/message/proto/v3/asset/store/ios_store/ios_store.proto';
import 'ostorlab/agent/message/proto/v3/asset/file/file.proto';
import 'ostorlab/agent/message/proto/v3/asset/file/android/aab/aab.proto';
import 'ostorlab/agent/message/proto/v3/asset/file/android/apk/apk.proto';
import 'ostorlab/agent/message/proto/v3/asset/file/ios/ipa/ipa.proto';

package ostorlab.agent.message.proto.v3.report.vulnerability;

enum RiskRating {
  HIGH = 0;
  MEDIUM = 1;
  LOW = 2;
  POTENTIALLY = 3;
  HARDENING = 4;
  SECURE = 5;
  IMPORTANT = 6;
  INFO = 7;
  CRITICAL = 8;
}

message Reference {
  optional string title = 1;
  optional string url = 2;
}
enum MetadataTypeEnum {
  FILE_PATH = 0;
  CODE_LOCATION = 1;
  URL = 2;
  PORT = 3;
  LOG = 4;
}
message Metadata {
  optional string value = 1;
  optional MetadataTypeEnum  type = 3;
}
message VulnerabilityLocation {
  oneof asset {
    ostorlab.agent.message.proto.v3.asset.domain_name.Message domain_name = 1;
    ostorlab.agent.message.proto.v3.asset.ip.v4.Message ipv4 = 2;
    ostorlab.agent.message.proto.v3.asset.ip.v6.Message ipv6 = 3;
    ostorlab.agent.message.proto.v3.asset.link.Message link = 4;
    ostorlab.agent.message.proto.v3.asset.store.android_store.Message android_store = 5;
    ostorlab.agent.message.proto.v3.asset.store.ios_store.Message ios_store = 6;
    ostorlab.agent.message.proto.v3.asset.file.Message file = 7;
    ostorlab.agent.message.proto.v3.asset.file.android.aab.Message android_aab = 8;
    ostorlab.agent.message.proto.v3.asset.file.android.apk.Message android_apk = 9;
    ostorlab.agent.message.proto.v3.asset.file.ios.ipa.Message ios_ipa = 10;
  }
  repeated Metadata metadata = 100 ;
}
message CategoryGroup {
  optional string key = 1;
  repeated string categories = 2;
}

message Message{
  optional string title = 1;
  optional RiskRating risk_rating = 2;
  optional string technical_detail = 3;
  optional string short_description = 4;
  optional string description = 5;
  optional string recommendation = 6;
  repeated Reference references = 7;
  optional bool security_issue = 8 [default = false];
  optional bool privacy_issue = 9 [default = false];
  optional bool has_public_exploit = 10 [default = false];
  optional bool targeted_by_malware = 11 [default = false];
  optional bool targeted_by_ransomware = 12 [default = false];
  optional bool targeted_by_nation_state = 13 [default = false];
  optional string dna = 14;
  optional string cvss_v3_vector = 15;
  optional VulnerabilityLocation vulnerability_location = 16;
  repeated CategoryGroup category_groups = 17;
  optional string cvss_v4_vector = 18;
  optional bool exploited = 19 [default = false];
  optional string exploitation_details = 20;
}
